pulsar-impl/src/main/java/com/datastax/oss/sink/pulsar/CassandraSinkTask.java

```diff
@@ -189,28 +199,22 @@ public void write(Record<T> record) throws Exception {
         log.info("raw {}", rawvalue);
       }
     }
-    int number;
-    synchronized (this) {
-      incomingList.add(record);
-      number = incomingList.size();
-    }
-    if (number == batchSize) {
+
+    incomingQueue.put(record); // blocking call to slow down the topic consumer
+    if (incomingQueue.size() >= batchSize) {
       flushExecutor.submit(this::flush);
     }
   }
 
   protected void flush() {
-    final List<Record<T>> swapList;
-    synchronized (this) {
-      if (!incomingList.isEmpty() && isFlushing.compareAndSet(false, true)) {
-        swapList = incomingList;
-        incomingList = new ArrayList<>();
-      } else {
-        return;
-      }
+    if (incomingQueue.isEmpty() || !isFlushing.compareAndSet(false, true)) {
+      return;
     }
+    final List<Record<T>> swapList = new ArrayList<>();
```

Comment:
```suggestion
    final List<Record<T>> swapList = new ArrayList<>(batchSize);
```
batch size or incomingQueue.size(), batchSize should be a good enough starting point

Reply:
I precisely ready `batchSize` as in line `214` instead of reading the entire queue. I think this lead to more predictability when interacting with the downstream and accurately matches the user preference of the batch size. 

metrics be handy here (to compare the queue size with the batch size at this point, which I suspect most of the time will be exactly the same) - will address those separately here #48